pkg/operator/cluster/ceph/mon/spec.go

```diff
@@ -55,47 +57,56 @@ func AdminSecretEnvVar() v1.EnvVar {
 	return v1.EnvVar{Name: "ROOK_ADMIN_SECRET", ValueFrom: &v1.EnvVarSource{SecretKeyRef: ref}}
 }
 
-func (c *Cluster) getLabels(name string) map[string]string {
+// PodNameEnvVar sets the Pod name as the name
+func PodNameEnvVar() v1.EnvVar {
+	return v1.EnvVar{Name: "ROOK_NAME", ValueFrom: &v1.EnvVarSource{
+		FieldRef: &v1.ObjectFieldSelector{FieldPath: "metadata.name"},
+	}}
+}
+
+func (c *Cluster) getLabels() map[string]string {
 	return map[string]string{
 		k8sutil.AppAttr: appName,
-		"mon":           name,
 		monClusterAttr:  c.Namespace,
 	}
 }
 
-func (c *Cluster) makeReplicaSet(config *monConfig, hostname string) *extensions.ReplicaSet {
-	rs := &extensions.ReplicaSet{
+func (c *Cluster) makeStatefulSet(replicas int32) *v1beta1.StatefulSet {
+	pod := c.makeMonPod()
+	sts := &v1beta1.StatefulSet{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:            config.Name,
+			Name:            appName,
 			Namespace:       c.Namespace,
 			OwnerReferences: []metav1.OwnerReference{c.ownerRef},
 		},
-	}
-
-	pod := c.makeMonPod(config, hostname)
-	replicaCount := int32(1)
-	rs.Spec = extensions.ReplicaSetSpec{
-		Template: v1.PodTemplateSpec{
-			ObjectMeta: pod.ObjectMeta,
-			Spec:       pod.Spec,
+		Spec: v1beta1.StatefulSetSpec{
+			UpdateStrategy: v1beta1.StatefulSetUpdateStrategy{
+				Type: v1beta1.RollingUpdateStatefulSetStrategyType,
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: pod.ObjectMeta,
+				Spec:       pod.Spec,
+			},
+			Replicas:    &replicas,
+			ServiceName: appName,
 		},
-		Replicas: &replicaCount,
 	}
 
-	return rs
+	return sts
 }
 
-func (c *Cluster) makeMonPod(config *monConfig, hostname string) *v1.Pod {
+func (c *Cluster) makeMonPod() *v1.Pod {
 	dataDirSource := v1.VolumeSource{EmptyDir: &v1.EmptyDirVolumeSource{}}
 	if c.dataDirHostPath != "" {
 		dataDirSource = v1.VolumeSource{HostPath: &v1.HostPathVolumeSource{Path: c.dataDirHostPath}}
 	}
 
-	container := c.monContainer(config, c.clusterInfo.FSID)
+	labels := c.getLabels()
+
+	container := c.monContainer(c.clusterInfo.FSID)
 	podSpec := v1.PodSpec{
 		Containers:    []v1.Container{container},
 		RestartPolicy: v1.RestartPolicyAlways,
-		NodeSelector:  map[string]string{apis.LabelHostname: hostname},
```

Comment:
without the nodeselector, a mon will not be tied to its host path. So if we failover the mon to a different node, it would have to rebuild the cache. I believe that would be ok if one mon fails. But what if multiple mons failover to different nodes? Seems like we have lost quorum and won't recover.